rts/motoko-rts/src/mark_stack.rs

```diff
@@ -0,0 +1,61 @@
+//! A stack for marking heap objects (for GC). There should be no allocation after the stack
+//! otherwise things will break as we push.
+
+use crate::alloc::{alloc_blob, alloc_words};
+use crate::types::{Blob, Words};
+
+/// Initial stack size
+const INIT_STACK_SIZE: Words<u32> = Words(64);
+
+/// Current stack
+static mut STACK_PTR: *mut Blob = core::ptr::null_mut();
+
+/// Current length
+static mut STACK_LEN: Words<u32> = Words(0);
+
+pub unsafe fn alloc_mark_stack() {
+    debug_assert!(STACK_PTR.is_null());
+
+    // Allocating an actual object here to not break dump_heap
+    STACK_PTR = alloc_blob(INIT_STACK_SIZE.to_bytes()).unskew() as *mut Blob;
+    STACK_LEN = Words(0);
+}
+
+pub unsafe fn free_mark_stack() {
+    STACK_PTR = core::ptr::null_mut();
+    STACK_LEN = Words(0);
+}
+
+/// Doubles the stack size
+unsafe fn grow_stack() {
+    let stack_cap: Words<u32> = STACK_PTR.len().to_words();
+    let p = alloc_words(stack_cap).unskew() as *mut usize;
+
+    // Make sure nothing was allocated after the stack
+    debug_assert_eq!(
+        (STACK_PTR.payload_addr() as *mut usize).add(STACK_LEN.0 as usize),
+        p
+    );
+
+    let new_cap: Words<u32> = Words(stack_cap.0 * 2);
+    (*STACK_PTR).len = new_cap.to_bytes();
+}
+
+pub unsafe fn push_mark_stack(obj: usize) {
+    if STACK_LEN == (*STACK_PTR).len.to_words() {
+        grow_stack();
+    }
+
+    *(STACK_PTR.payload_addr() as *mut usize).add(STACK_LEN.0 as usize) = obj;
+    STACK_LEN += Words(1);
+}
+
+pub unsafe fn pop_mark_stack() -> Option<usize> {
```

Comment:
I should try to get rid of this `Option` as well, I think it could also improve perf. (https://github.com/dfinity/motoko/pull/2522/files#r649170651)

Reply:
Cool again

Reply:
I don't understand why but if I apply the same technique here the benchmark starts using more instructions. No idea why. The new `pop_mark_stack` is actually smaller in generated Wasm. This is the new version I've tried:

```rust
pub unsafe fn pop_mark_stack() -> usize {
    if STACK_LEN.0 == 0 {
        0
    } else {
        STACK_LEN -= Words(1);
        *(STACK_PTR.payload_addr() as *mut usize).add(STACK_LEN.0 as usize)
    }
}
```

Reply:
Here's the patch for this change:

```diff
diff --git a/rts/motoko-rts-tests/src/mark_stack.rs b/rts/motoko-rts-tests/src/mark_stack.rs
index 0499a31bf..4938f252a 100644
--- a/rts/motoko-rts-tests/src/mark_stack.rs
+++ b/rts/motoko-rts-tests/src/mark_stack.rs
@@ -24,7 +24,7 @@ fn test_(objs: Vec<usize>) -> TestResult {
 
         for obj in objs.iter().rev() {
             let popped = pop_mark_stack();
-            if popped != Some(*obj) {
+            if popped != *obj {
                 free_mark_stack(); // TODO: Does not really free
                 return TestResult::error(format!(
                     "Unexpected object popped, expected={:?}, popped={:?}",
diff --git a/rts/motoko-rts/src/gc/mark_compact.rs b/rts/motoko-rts/src/gc/mark_compact.rs
index ef8a814ee..5890d67db 100644
--- a/rts/motoko-rts/src/gc/mark_compact.rs
+++ b/rts/motoko-rts/src/gc/mark_compact.rs
@@ -80,7 +80,11 @@ unsafe fn push_mark_stack(obj: SkewedPtr, heap_base: u32) {
 }
 
 unsafe fn mark_stack(heap_base: u32) {
-    while let Some(obj) = pop_mark_stack() {
+    loop {
+        let obj = pop_mark_stack();
+        if obj == 0 {
+            break;
+        }
         mark_fields(obj as *mut Obj, heap_base);
     }
 }
diff --git a/rts/motoko-rts/src/mark_stack.rs b/rts/motoko-rts/src/mark_stack.rs
index be87acc6f..1bf8145b9 100644
--- a/rts/motoko-rts/src/mark_stack.rs
+++ b/rts/motoko-rts/src/mark_stack.rs
@@ -50,12 +50,11 @@ pub unsafe fn push_mark_stack(obj: usize) {
     STACK_LEN += Words(1);
 }
 
-pub unsafe fn pop_mark_stack() -> Option<usize> {
+pub unsafe fn pop_mark_stack() -> usize {
     if STACK_LEN.0 == 0 {
-        None
+        0
     } else {
         STACK_LEN -= Words(1);
-        let p = *(STACK_PTR.payload_addr() as *mut usize).add(STACK_LEN.0 as usize);
-        Some(p)
+        *(STACK_PTR.payload_addr() as *mut usize).add(STACK_LEN.0 as usize)
     }
 }
```

Reply:
Any idea why the cost went up?

Reply:
Here's the diff of before vs. after this patch (mo-rts.wasm):

```diff
--- mo-rts.wat  2021-06-14 15:44:34.892004845 +0300
+++ ../../motoko/rts/mo-rts.wat 2021-06-14 15:44:34.180013747 +0300
@@ -19343,7 +19343,7 @@
     local.get 5
     i32.const 1
     i32.add
-    local.set 10
+    local.set 8
     block  ;; label = @1
       local.get 5
       i32.const 5
@@ -19352,7 +19352,7 @@
       local.tee 2
       i32.eqz
       br_if 0 (;@1;)
-      local.get 10
+      local.get 8
       i32.const 8
       i32.add
       local.set 1
@@ -19402,7 +19402,6 @@
         local.tee 2
         i32.const 15796
         i32.add
-        local.tee 8
         local.get 1
         i32.const -1
         i32.add
@@ -19419,21 +19418,27 @@
         i32.const 8
         i32.add
         i32.load
+        local.tee 1
+        i32.eqz
+        br_if 1 (;@1;)
+        local.get 1
         local.get 4
         call $motoko_rts::gc::mark_compact::mark_fields::hc1bca295efbdeb41
-        local.get 8
+        global.get 1
+        i32.const 15796
+        i32.add
         i32.load
         local.tee 1
         br_if 0 (;@2;)
       end
     end
     block  ;; label = @1
-      local.get 10
+      local.get 8
       i32.load offset=4 align=1
       local.tee 2
       i32.eqz
       br_if 0 (;@1;)
-      local.get 10
+      local.get 8
       i32.const 8
       i32.add
       local.set 1
```

So the new version, after a lot of inlining, ends up being larger. No idea why though. This code is for `compacting_gc`, so `mark_stack` is inlined in `mark_compact`, which is inlined in `compacting_gc`. Somehow the version with this change has more stack traffic.

Reply:
Maybe it would be shorter if you just stored the sentinel value 0 at the bottom of the stack and avoided doing the conditional testing on pop?

Reply:
And maybe just maintain pointer to top of stack rather than computing from length on push and pop

Reply:
> And maybe just maintain pointer to top of stack rather than computing from length on push and pop

This made a benchmark use 2.3% less instructions. Pushed as 836992cfc.

Reply:
> Maybe it would be shorter if you just stored the sentinel value 0 at the bottom of the stack and avoided doing the conditional testing on pop?

We would need a conditional to avoid popping the sentinel value, right? Otherwise a push after popping the sentinel value would break the stack. So I think we can't remove the conditional in `pop_mark_stack`.